Add gtk3 and its missing dependencies.

ryanvolz opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with
a PR or once it's ready for review, please let the right people know.
There are language-specific teams for reviewing recipes.

Currently available teams are:
- python `@conda-forge/help-python`
- python/c hybrid `@conda-forge/help-python-c`
- r `@conda-forge/help-r`
- java `@conda-forge/help-java`
- nodejs `@conda-forge/help-nodejs`
- c/c++ `@conda-forge/help-c-cpp`
- perl `@conda-forge/help-perl`
- Julia `@conda-forge/help-julia`
- ruby `@conda-forge/help-ruby`

If your PR doesn't fall into those categories please contact
the full review team `@conda-forge/staged-recipes`.

Due to GitHub limitations first time contributors to conda-forge are unable
to ping these teams. You can [ping the team](https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team) using a special command in 
a comment on the PR to get the attention of the `staged-recipes` team. You can 
also consider asking on our [Gitter channel](https://gitter.im/conda-forge/conda-forge.github.io)
if your recipe isn't reviewed promptly.
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml" 
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

Fulfills package request #9314. This is a big name package, so I expect discussion and just wanted to get the ball rolling. My personal interest is in enabling use of gnuradio-companion (https://github.com/conda-forge/gnuradio-feedstock/issues/4).

These recipes are adapted from the recipes by @pkgw found at https://github.com/pkgw/conda-recipes. Other than updating them to the most recent versions and getting the build to work on my local Linux machine, the main changes include dropping Wayland support (for now, I wanted to start with a minimal package to get it merged) and depending on CDTs for the X11 dependencies. https://github.com/conda-forge/cfep/pull/7 is relevant to the latter decision, but pending any movement on that, I thought it might be best to follow what has already worked for the Qt recipe.

I welcome anyone who else who wants to take on maintainership (@pkgw?).

I've also never tried merging multiple packages at once, so we'll see how that goes.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/at-spi2-atk```, ```recipes/at-spi2-core```, ```recipes/epoxy```, ```recipes/gtk3```) and found it was in an excellent condition.

stuarteberg commented:
In the linux build, the current failure is:

```
  ERROR (at-spi2-core,lib/libatspi.so.0.0.1): Needed DSO lib/libX11.so.6 found in ['xorg-libx11']
  ERROR (at-spi2-core,lib/libatspi.so.0.0.1): .. but ['xorg-libx11'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
```

I'm not an expert, but according to [this discussion][1], the answer is to add `xorg-libx11` to the `run` dependencies.

[1]: https://gitter.im/conda-forge/conda-forge.github.io?at=5e43c3b81d23aa47aa0c23dc

On Mac, the errors are different (and more numerous).  Haven't investigated those yet.

ryanvolz commented:
Yes, this is stemming from the decision to depend on the CDTs instead of the xorg- packages, but some of the xorg packages are pulled in by other dependencies (cairo) and linked against anyway. I'm testing to see if it would work to remove the libraries installed by xorg-x11 before building.

The OSX error so far is just that it found an optional dependency in ``gettext`` but that wasn't in the dependency list. That one is easy to fix.

ryanvolz commented:
I'm a bit stumped at why the OSX build is failing at an ``install_name_tool`` command, and I don't have a local machine to test. Not much to go on from the Azure log:
```
Installing gtk/gtk-query-settings to $PREFIX/bin
Command '['install_name_tool', '/usr/local/miniconda/conda-bld/gtk3_1583274843921/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/bin/gtk-query-settings', '-add_rpath', '/usr/local/miniconda/conda-bld/gtk3_1583274843921/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib', '-change', '@rpath/libgdk-3.0.dylib', '/usr/local/miniconda/conda-bld/gtk3_1583274843921/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/libgdk-3.0.dylib', '-change', '@rpath/libgtk-3.0.dylib', '/usr/local/miniconda/conda-bld/gtk3_1583274843921/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/libgtk-3.0.dylib']' returned non-zero exit status 1.
FAILED: meson-install
```
Any ideas, @conda-forge/staged-recipes? Other than that, the Linux build is working and this is ready for feedback.

stuarteberg commented:
>I'm a bit stumped at why the OSX build is failing at an install_name_tool command, and I don't have a local machine to test.

I gave it a try on my MacBook.  The failing command and error message are shown below (with edits for readability):

```
install_name_tool \
    /opt/miniconda/conda-bld/gtk3_1583298399113/_h_env_placehold/bin/gtk-query-settings \
    -add_rpath /opt/miniconda/conda-bld/gtk3_1583298399113/_h_env_placehold/lib \
    -change @rpath/libgdk-3.0.dylib /opt/miniconda/conda-bld/gtk3_1583298399113/_h_env_placehold/lib/libgdk-3.0.dylib \
    -change @rpath/libgtk-3.0.dylib /opt/miniconda/conda-bld/gtk3_1583298399113/_h_env_placehold/lib/libgtk-3.0.dylib

error: 
  .../install_name_tool:
  for: /opt/miniconda/conda-bld/gtk3_1583298399113/_h_env_placehold/bin/gtk-query-settings (for architecture x86_64)
  option "-add_rpath /opt/miniconda/conda-bld/gtk3_1583298399113/_h_env_placehold/lib"
  would duplicate path, file already has LC_RPATH
  for: /opt/miniconda/conda-bld/gtk3_1583298399113/_h_env_placehold/lib
```

So it's complaining that conda-build's post-processing logic is attempting to add an `rpath` to a binary (`gtk-query-settings`) that alread contains the exact same `rpath`.

Could `recipes/gtk3/meson-rpaths.patch` be to blame?  Should that patch be selectively applied (i.e. linux only)?

(TBH, I'm surprised.  I thought that conda's post-link logic refrained from adding `rpaths` to things that don't need them.)

isuruf commented:
Try removing `recipes/gtk3/meson-rpaths.patch`. `LDFLAGS` env variable already has `rpath`, so if meson also tries to add the rpath, it will fail.

stuarteberg commented:
Indeed, the build passes on my MacBook if I remove `recipes/gtk3/meson-rpaths.patch`.  If it's really needed on Linux for some reason, then add a selector (`# [linux]`) to prevent it from being used on Mac.

ryanvolz commented:
> I don't know why `meson-rpaths.patch` was needed in the first place, but it certainly does look like it's not needed now. Good catch, @stuarteberg!

This is why the patch exists, and it is still needed on Linux: https://github.com/mesonbuild/meson/issues/4685. I added a comment to the ``meta.yaml`` to that effect.

ryanvolz commented:
> > I don't know why `meson-rpaths.patch` was needed in the first place, but it certainly does look like it's not needed now. Good catch, @stuarteberg!
> 
> This is why the patch exists, and it is still needed on Linux: [mesonbuild/meson#4685](https://github.com/mesonbuild/meson/issues/4685). I added a comment to the `meta.yaml` to that effect.

This was true, but now the ``meson`` package in ``conda-forge`` has a patch (https://github.com/conda-forge/meson-feedstock/commit/777a4844a31095ca20a5b9b7545e99156569189d) that works around the linked issue, so the ``meson-rpaths.patch`` is not actually necessary. The Linux builds are now failing for an as-yet unknown reason, despite the fact that nothing should have changed since they succeeded yesterday.

ryanvolz commented:
> The Linux builds are now failing for an as-yet unknown reason, despite the fact that nothing should have changed since they succeeded yesterday.

I can confirm from local testing that the Linux builds are now failing as a result of the upgrade from ``conda-build=3.18.11`` to ``conda-build=3.18.12``. Specifically, I'm now seeing 
```
gtk-launch: symbol lookup error: gtk-launch: undefined symbol: gtk_get_minor_version
```
when trying to run `gtk-launch` during the testing phase when it ran successfully previously. Any ideas about what specific change might be causing this, @conda-forge/staged-recipes?

stuarteberg commented:
I am also seeing some weird behavior in a different feedstock.  I've been trying to pin down the cause, and the new `conda-build` release is something I was wondering about.

stuarteberg commented:
I'm trying to figure out which change in `conda-build 3.18.12` could be the culprit (assuming the `conda-build` release is the cause).  So far, the following change sets seem worth investigating:

- https://github.com/conda/conda-build/pull/3791 (touches the logic that edits hard-coded prefixes in binary files)
- https://github.com/conda/conda-build/pull/3808 (touches the logic that updates RPATHs, etc.)

Both were written by @mingwandroid, since he is a wizard when it comes to this stuff. 😄 

So, Ray, here's the situation: The linux build for this recipe was passing yesterday, but after upgrading conda-build, it now fails (see error above).  Do you see anything obviously wrong with this recipe that might cause issues in the new conda-build, but not the old version?

mingwandroid commented:
Looking now..

mingwandroid commented:
Very, very cool to see gtk3 being worked on BTW!

ryanvolz commented:
If it helps, here's a diff on the result of `objdump -x gtk-launch` from the working version with `conda-build 3.18.11` to the failing version with `conda-build 3.18.12`:
```
@@ -38,7 +38,7 @@
   NEEDED               libpthread.so.0
   NEEDED               libc.so.6
   SYMBOLIC             0x0000000000000000
-  RPATH                $ORIGIN/../lib:$ORIGIN/:$ORIGIN/../gdk
+  RPATH                $ORIGIN/../lib
   INIT                 0x0000000000002000
   FINI                 0x0000000000002738
   HASH                 0x00000000000002e8
@@ -79,7 +79,7 @@
                   CONTENTS, ALLOC, LOAD, READONLY, DATA
   4 .dynsym       00000558  0000000000000608  0000000000000608  00000608  2**3
                   CONTENTS, ALLOC, LOAD, READONLY, DATA
-  5 .dynstr       000006f4  0000000000000b60  0000000000000b60  00000b60  2**0
+  5 .dynstr       00000418  0000000000000b60  0000000000000b60  00000b60  2**0
                   CONTENTS, ALLOC, LOAD, READONLY, DATA
   6 .gnu.version  00000072  0000000000001254  0000000000001254  00001254  2**1
                   CONTENTS, ALLOC, LOAD, READONLY, DATA
```
And the output of `readelf -p .dynstr` for `conda-build 3.18.11`:
```
String dump of section '.dynstr':
  [     1]  libgtk-3.so.0
  [     f]  g_bus_get_sync
  [    1e]  g_option_context_parse
  [    35]  g_option_context_add_group
  [    50]  g_file_new_for_commandline_arg
  [    6f]  g_error_free
  [    7c]  g_type_check_instance_cast
  [    97]  g_option_context_new
  [    ac]  g_printerr
  [    b7]  g_list_append
  [    c5]  g_strdup
  [    ce]  g_get_prgname
  [    dc]  g_strdup_printf
  [    ec]  g_desktop_app_info_new
  [   103]  gdk_display_get_default
  [   11b]  g_app_info_launch
  [   12d]  g_dbus_connection_call_sync
  [   149]  g_app_info_get_type
  [   15d]  g_option_context_free
  [   173]  g_list_free_full
  [   184]  g_free
  [   18b]  g_strconcat
  [   197]  g_option_context_add_main_entries
  [   1b9]  g_object_unref
  [   1c8]  gdk_display_get_app_launch_context
  [   1eb]  g_str_has_suffix
  [   1fc]  _ITM_deregisterTMCloneTable
  [   218]  __gmon_start__
  [   227]  _ITM_registerTMCloneTable
  [   241]  gtk_get_option_group
  [   256]  gtk_get_minor_version
  [   26c]  gtk_init
  [   275]  gtk_get_micro_version
  [   28b]  gtk_get_major_version
  [   2a1]  libgdk-3.so.0
  [   2af]  libglib-2.0.so.0
  [   2c0]  g_print
  [   2c8]  g_option_context_set_summary
  [   2e5]  libgobject-2.0.so.0
  [   2f9]  libgio-2.0.so.0
  [   309]  g_dbus_is_name
  [   318]  libpthread.so.0
  [   328]  libc.so.6
  [   332]  setlocale
  [   33c]  dcgettext
  [   346]  __stack_chk_fail
  [   357]  bind_textdomain_codeset
  [   36f]  __cxa_finalize
  [   37e]  bindtextdomain
  [   38d]  __libc_start_main
  [   39f]  _edata
  [   3a6]  __bss_start
  [   3b2]  _end
  [   3b7]  __libc_csu_fini
  [   3c7]  __data_start
  [   3d4]  __libc_csu_init
  [   3e4]  _IO_stdin_used
  [   3f3]  GLIBC_2.4
  [   3fd]  GLIBC_2.2.5
  [   409]  $ORIGIN/../lib:$ORIGIN/:$ORIGIN/../gdk
  [   430]  XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
```
And the output of `readelf -p .dynstr` for `conda-build 3.18.12`:
```
String dump of section '.dynstr':
  [     1]  libgtk-3.so.0
  [     f]  libgdk-3.so.0
  [    1d]  libglib-2.0.so.0
  [    2e]  libgobject-2.0.so.0
  [    42]  libgio-2.0.so.0
  [    52]  libpthread.so.0
  [    62]  libc.so.6
  [    6c]  $ORIGIN/../lib
  [    7b]  gdk_display_get_app_launch_context
  [    9e]  g_option_context_add_main_entries
  [    c0]  g_file_new_for_commandline_arg
  [    df]  g_option_context_set_summary
  [    fc]  g_dbus_connection_call_sync
  [   118]  _ITM_deregisterTMCloneTable
  [   134]  g_option_context_add_group
  [   14f]  g_type_check_instance_cast
  [   16a]  _ITM_registerTMCloneTable
  [   184]  gdk_display_get_default
  [   19c]  bind_textdomain_codeset
  [   1b4]  g_desktop_app_info_new
  [   1cb]  g_option_context_parse
  [   1e2]  gtk_get_major_version
  [   1f8]  gtk_get_micro_version
  [   20e]  gtk_get_minor_version
  [   224]  g_option_context_free
  [   23a]  gtk_get_option_group
  [   24f]  g_option_context_new
  [   264]  g_app_info_get_type
  [   278]  __libc_start_main
  [   28a]  g_app_info_launch
  [   29c]  g_list_free_full
  [   2ad]  g_str_has_suffix
  [   2be]  __stack_chk_fail
  [   2cf]  __libc_csu_fini
  [   2df]  __libc_csu_init
  [   2ef]  g_strdup_printf
  [   2ff]  __cxa_finalize
  [   30e]  _IO_stdin_used
  [   31d]  g_object_unref
  [   32c]  g_dbus_is_name
  [   33b]  __gmon_start__
  [   34a]  bindtextdomain
  [   359]  g_bus_get_sync
  [   368]  g_get_prgname
  [   376]  g_list_append
  [   384]  __data_start
  [   391]  g_error_free
  [   39e]  __bss_start
  [   3aa]  g_strconcat
  [   3b6]  g_printerr
  [   3c1]  setlocale
  [   3cb]  dcgettext
  [   3d5]  g_strdup
  [   3de]  gtk_init
  [   3e7]  g_print
  [   3ef]  _edata
  [   3f6]  g_free
  [   3fd]  _end
  [   402]  GLIBC_2.4
  [   40c]  GLIBC_2.2.5
```

ryanvolz commented:
I found another clue: `gtk_get_minor_version` is not in the dynamic symbol table of `libgtk-3.so` (`nm -D libgtk-3.so`) with `conda-build 3.18.12`, but it is with `conda-build 3.18.11`. In fact, seemingly all of the `gtk_*` symbols are missing.

This is getting pretty arcane for me, but I'll continue digging tomorrow if no one else makes headway in the meantime.

mingwandroid commented:
I'm going to have to go to sleep soon. I'll fire up my linux box tomorrow though, WSL isn't cutting it.

mingwandroid commented:
WSL build completed but failed on test due to me not installing system libs:

```
+ gtk-encode-symbolic-svg --help
gtk-encode-symbolic-svg: error while loading shared libraries: libXcursor.so.1: cannot open shared object file: No such file or directory
```

Test packages on `rdonnelly`. Going to call it a night.

ryanvolz commented:
This looks like it might be a bug with LIEF? I built gtk3 locally, went to the work directory, and found the `libgtk-3.so` as produced directly by the build. It has all of the `gtk_*` symbols in the dynamic symbol table.
```
$ nm -D libgtk-3.so | grep gtk_get_minor_version
0000000000200970 T gtk_get_minor_version
```
If I then open up that library with LIEF and list the dynamic symbols from there,
```
In [1]: import lief                                                              

In [2]: binary = lief.parse("libgtk-3.so")                              

In [3]: dsyms = [s.demangled_name for s in binary.dynamic_symbols]               

In [4]: 'gtk_get_minor_version' in dsyms                                         
False
```
none of the `gtk_*` symbols are included. Simply writing the library back out with LIEF
```
In [5]: binary.write("libgtk-3.so.modified")
```
produces a library that looks much like the output with `conda-build 3.18.12`:
```
$ nm -D libgtk-3.so.modified | grep gtk_get_minor_version
(no output)
```
I haven't had time to grok the changes with `conda-build 3.18.12`, but I'd guess it's now passing the binaries through LIEF when it didn't necessarily do that before. Does that make sense, @mingwandroid?

ryanvolz commented:
Looks like the bug is known: https://github.com/lief-project/LIEF/issues/239

Using the workaround from the above
```
binary = lief.ELF.parse("libgtk-3.so", lief.ELF.DYNSYM_COUNT_METHODS.SECTION)
```
fixes my test case from the previous comment. Perhaps something similar can be put in conda-build?

isuruf commented:
You can set `build: rpaths_patcher: patchelf` to force patchelf.

ryanvolz commented:
> You can set `build: rpaths_patcher: patchelf` to force patchelf.

Great! I'll give this a try.

stuarteberg commented:
Another confusing thing is that the log included warnings like the following (edited for clarity)

```
2020-03-04T21:30:05.6338173Z WARNING :: get_rpaths_raw()=[
'/home/conda/staged-recipes/build_artifacts/at-spi2-core_1583357277553/_build_env/lib
:/home/conda/staged-recipes/build_artifacts/at-spi2-core_1583357277553/_h_env_placehold/lib
:/home/conda/staged-recipes/build_artifacts/at-spi2-core_1583357277553/_h_env_placehold/lib
:/home/conda/staged-recipes/build_artifacts/at-spi2-core_1583357277553/_build_env/bin/../x86_64-conda_cos6-linux-gnu/sysroot/usr/lib64']

and patchelf=[[
'/home/conda/staged-recipes/build_artifacts/at-spi2-core_1583357277553/_build_env/lib',
'/home/conda/staged-recipes/build_artifacts/at-spi2-core_1583357277553/_h_env_placehold/lib',
'/home/conda/staged-recipes/build_artifacts/at-spi2-core_1583357277553/_h_env_placehold/lib', 
'/home/conda/staged-recipes/build_artifacts/at-spi2-core_1583357277553/_build_env/bin/../x86_64-conda_cos6-linux-gnu/sysroot/usr/lib64']]

disagree for /home/conda/staged-recipes/build_artifacts/at-spi2-core_1583357277553/_h_env_placehold/libexec/at-spi-bus-launcher :: 
```

There isn't actually any disagreement, but `get_rpaths_raw()` is returning the RPATH a single string with `:` separators, instead of a list of directories. (I added the newlines.)

ryanvolz commented:
Build success, green check, ready for review @conda-forge/staged-recipes!

mingwandroid commented:
Urgh, sorry everyone, partchelf should have been the default here. I'll work on making this robust.

The problem is both are buggy! I'm hoping lief 0.10.1 is less so. It's nearly ready too.

ryanvolz commented on recipes/at-spi2-core/no-intltool.patch:
```diff
@@ -0,0 +1,33 @@
++fi
+ 
+ if test -z "$AM_DEFAULT_VERBOSITY"; then
+   AM_DEFAULT_VERBOSITY=1
```
No, it looks like it's a holdover from the original recipe where it was only partially removed a couple years ago.

ryanvolz commented:
> But you also pointed out that one of the dependencies (`cairo`) depends on conda-forge's own x11 packages. So the `libX11.so` is actually present in `$PREFIX/lib` at build time, right. Are you sure this build picked up the system version of `libX11.so` and not the conda-forge version? Does it matter?

I'm no expert either, but I don't think it matters (at least it doesn't seem to). The build does pick up the `libX11.so` from conda-forge's own xorg packages pulled in by `cairo`, and then it will use those same libraries again at runtime because of `cairo`, but the ABI is seemingly stable enough that `cairo` doesn't even have to constrain the runtime dependency against a particular version. `gtk3` does use the CDTs for X11 libraries that it needs outside of what `cairo` brings in, and still this mixing seems to work without ill effect. My thinking is that, were `cairo` to change its X11 dependencies, this would still work fine and it puts `gtk3` on level footing with how the `qt` package handles X11.

It would be good, though, if someone who knows better could weigh in. It would be a pretty simple change to just depend on conda-forge's own xorg packages, and maybe that's the cleaner thing to do.

stuarteberg commented:
>It would be good, though, if someone who knows better could weigh in.

OK, I've asked for advice on the conda-forge gitter channel:
https://gitter.im/conda-forge/conda-forge.github.io?at=5e62703eafdc1e2033d3ef22

isuruf commented:
> My thinking is that, were cairo to change its X11 dependencies, this would still work fine and it puts gtk3 on level footing with how the qt package handles X11.

Couldn't have said it better myself.

isuruf commented on recipes/epoxy/build.sh:
```diff
@@ -0,0 +1,36 @@
+    # downstream deps don't all have to include a magic CDT package as a dependency
+    mkdir -p $PREFIX/include/KHR $PREFIX/include/EGL
+    cp $BUILD_PREFIX/$HOST/sysroot/usr/include/KHR/khrplatform.h $PREFIX/include/KHR/
+    cp $BUILD_PREFIX/$HOST/sysroot/usr/include/EGL/eglplatform.h $PREFIX/include/EGL/
```
I'm not sure the 3 lines above are desirable.

ryanvolz replied:
Hmm, fair. Considering that "downstream deps" is just `gtk3` at this point, hackery is probably not warranted.

stuarteberg commented on recipes/epoxy/build.sh:
```diff
@@ -0,0 +1,27 @@
+
+set -ex
+
+if [ -n "$OSX_ARCH" ] ; then
```
FWIW, in other staged-recipe PRs, I've seen the conda-forge core team ask people to use the following method to detect OSX.  I don't know if they will request it here, too.

```bash
if [[ $target_platform == osx* ]]
```

isuruf replied:
That's a good suggestion.

ryanvolz replied:
Oooh, I do like that better.

stuarteberg commented:
Above, you mentioned that @pkgw would be a natural person to invite as a maintainer.  What do you think @pkgw?

If you want more maintainers, feel free to add me as well, though I don't have the same level of expertise.

Otherwise, FWIW, this recipe looks good to me.  (I'm not a conda-forge admin, though.)  Thanks for getting this building!  I'm looking forward to using it.

stuarteberg commented:
@conda-forge/help-c-cpp @isuruf Can this be merged?  Unless @ryanvolz thinks there's anything left to do here, I think it's ready to go.  (We can always add more maintainers later.)

ryanvolz commented:
It's ready as far as I'm concerned.

On March 15, 2020 1:55:41 PM EDT, Stuart Berg <notifications@github.com> wrote:
>@conda-forge/help-c-cpp @isuruf Can this be merged?  Unless @ryanvolz
>thinks there's anything left to do here, I think it's ready to go.  (We
>can always add more maintainers later.)
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/conda-forge/staged-recipes/pull/10979#issuecomment-599243087

isuruf commented on recipes/at-spi2-core/meta.yaml:
```diff
@@ -0,0 +1,64 @@
+  run_exports:
+    - {{ pin_subpackage('at-spi2-core', max_pin='x.x') }}
+  missing_dso_whitelist:  # [linux]
+    - '*/libX11.so*'  # [linux]
```
Is the above needed if we are using CDTs?

ryanvolz replied:
The issue is that because `xorg-x11` is installed through the `cairo` dependency, the DSO checking picks up that the non-CDT library is linked in but that it is not present as a run requirement, so the check without fail without something like this.

isuruf replied:
Ah, okay

isuruf commented on recipes/epoxy/build.sh:
```diff
@@ -0,0 +1,31 @@
+    --prefix=$PREFIX \
+    --libdir=$PREFIX/lib  \
+    --wrap-mode=nofallback
+ninja -v -C builddir
```
```suggestion
ninja -v -C builddir -j ${CPU_COUNT}
```

isuruf commented on recipes/epoxy/build.sh:
```diff
@@ -0,0 +1,31 @@
+    --libdir=$PREFIX/lib  \
+    --wrap-mode=nofallback
+ninja -v -C builddir
+ninja -C builddir install
```
```suggestion
ninja -C builddir install -j ${CPU_COUNT}
```

isuruf commented on recipes/gtk3/build.sh:
```diff
@@ -0,0 +1,32 @@
+    --libdir=$PREFIX/lib  \
+    --wrap-mode=nofallback
+ninja -v -C builddir
+ninja -C builddir install
```
Same here with CPU_COUNT

isuruf commented on recipes/gtk3/meta.yaml:
```diff
@@ -0,0 +1,132 @@
+    - {{ pin_subpackage('gtk3', max_pin='x') }}
+  missing_dso_whitelist:
+    - '*/libX11.so*'  # [linux]
+    - '*/libXext.so*'  # [linux]
```
Same here

isuruf commented on recipes/at-spi2-atk/build.sh:
```diff
@@ -0,0 +1,17 @@
+    --libdir=$PREFIX/lib  \
+    --wrap-mode=nofallback
+ninja -v -C builddir
+ninja -C builddir install
```
Same here

ryanvolz commented:
@isuruf's most recent comments have now been Incorporated.

stuarteberg commented:
Thanks @isuruf!

stuarteberg commented:
The feedstock build failed at first since they all tried to build at once.  (Gtk3 failed because the dependencies weren't ready yet.)  But I was able to restart those builds, and now they're all there.  `conda install -c conda-forge gtk3` now works!

ryanvolz commented:
Thanks for the help and support everyone! I'll also note here that I'm a good chunk of the way to getting a Windows build working; just have to send out some pull requests to existing feedstocks.
